File: app/src/main/java/com/tommihirvonen/exifnotes/activities/FramesActivity.java
package com.tommihirvonen.exifnotes.activities;

import android.content.Intent;
import android.os.Bundle;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.tommihirvonen.exifnotes.R;
import com.tommihirvonen.exifnotes.fragments.FramesFragment;
import com.tommihirvonen.exifnotes.utilities.ExtraKeys;
import com.tommihirvonen.exifnotes.utilities.Utilities;

/**
 * Activity to contain the fragment for frames
 */
public class FramesActivity extends AppCompatActivity {

    /**
     * Constants passed to PreferenceActivity
     */
    public static final int PREFERENCE_ACTIVITY_REQUEST = 8;

    @Override
    protected void onCreate(@Nullable final Bundle savedInstanceState) {

        // Get the arguments from the intent from MainActivity.
        final Intent intent = getIntent();
        final long rollId = intent.getLongExtra(ExtraKeys.ROLL_ID, -1);
        final boolean locationEnabled = intent.getBooleanExtra(ExtraKeys.LOCATION_ENABLED, false);
        final boolean overridePendingTransition = intent.getBooleanExtra(ExtraKeys.OVERRIDE_PENDING_TRANSITION, false);
        if (rollId == -1) finish();

        // If the activity was launched from MainActivity, enable custom transition animations.
        // If the activity was recreated, then custom animations are not enabled during onCreate().
        if (overridePendingTransition) {
            overridePendingTransition(R.anim.enter_from_right, R.anim.hold);
            // Replace the old intent with a modified one, where the OVERRIDE_PENDING_TRANSITION
            // boolean value has been exhausted and set to false.
            intent.putExtra(ExtraKeys.OVERRIDE_PENDING_TRANSITION, false);
            setIntent(intent);
        }

        if (Utilities.isAppThemeDark(getBaseContext())) setTheme(R.style.AppTheme_Dark);

        // The point at which super.onCreate() is called is important.
        // Calling it at the end of the method resulted in the back button not appearing
        // when action mode was enabled.
        super.onCreate(savedInstanceState);

        // Use the same activity layout as in MainActivity.
        setContentView(R.layout.activity_main);

        Utilities.setUiColor(this, true);

        if (findViewById(R.id.fragment_container) != null && savedInstanceState == null) {

            // Pass the arguments from MainActivity on to FramesFragment.
            final FramesFragment framesFragment = new FramesFragment();
            final Bundle arguments = new Bundle();
            arguments.putLong(ExtraKeys.ROLL_ID, rollId);
            arguments.putBoolean(ExtraKeys.LOCATION_ENABLED, locationEnabled);
            framesFragment.setArguments(arguments);

            getSupportFragmentManager().beginTransaction().add(R.id.fragment_container,
                    framesFragment, FramesFragment.FRAMES_FRAGMENT_TAG).commit();

            // Bring the shadow element from the activity's layout to front.
            findViewById(R.id.shadow).bringToFront();

        }

    }

    @Override
    protected void onResume() {
        super.onResume();
        final int primaryColor = Utilities.getPrimaryUiColor(getBaseContext());
        final int secondaryColor = Utilities.getSecondaryUiColor(getBaseContext());
        Utilities.setSupportActionBarColor(this, primaryColor);
        Utilities.setStatusBarColor(this, secondaryColor);
    }

    @Override
    public void finish() {
        super.finish();
        overridePendingTransition(R.anim.nothing, R.anim.exit_to_right);
    }

    /**
     * The PreferenceActivity is started for result and the result is captured here.
     *
     * The result code is compared using bitwise operators to determine
     * whether a new database was imported, the app theme was changed or both.
     *
     * @param requestCode passed to the activity when it is launched
     * @param resultCode integer to be compared using bitwise operators to determine the action(s)
     *                   that were taken in PreferenceActivity
     * @param data not used
     */
    @Override
    public void onActivityResult(final int requestCode, final int resultCode, final Intent data) {
        // If a new database was imported, use setResult() to notify MainActivity as well.
        if (requestCode == PREFERENCE_ACTIVITY_REQUEST) {
            // Finish this activity since the selected roll may not be valid anymore.
            if ((resultCode & PreferenceActivity.RESULT_DATABASE_IMPORTED) ==
                    PreferenceActivity.RESULT_DATABASE_IMPORTED) {
                setResult(resultCode);
                finish();
            }
            // If the app theme was changed, recreate activity for changes to take effect.
            if ((resultCode & PreferenceActivity.RESULT_THEME_CHANGED) ==
                    PreferenceActivity.RESULT_THEME_CHANGED) {
                recreate();
            }
            return;
        }

        // Call super in case the result was not handled here
        super.onActivityResult(requestCode, resultCode, data);
    }

}

File: app/src/main/java/com/tommihirvonen/exifnotes/activities/FramesActivity.kt
/*
 * Exif Notes
 * Copyright (C) 2022  Tommi Hirvonen
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.tommihirvonen.exifnotes.activities

import android.Manifest
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.view.View
import android.view.WindowManager
import android.widget.FrameLayout
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.preference.PreferenceManager
import com.tommihirvonen.exifnotes.R
import com.tommihirvonen.exifnotes.datastructures.Roll
import com.tommihirvonen.exifnotes.fragments.FramesFragment
import com.tommihirvonen.exifnotes.preferences.PreferenceConstants
import com.tommihirvonen.exifnotes.utilities.*

/**
 * Activity to contain the fragment for frames
 */
class FramesActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        val prefs = PreferenceManager.getDefaultSharedPreferences(this)
        val requestingLocationUpdates = prefs.getBoolean(PreferenceConstants.KEY_GPS_UPDATE, true)

        val permissionAccessCoarseLocation = ActivityCompat.checkSelfPermission(
            this, Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED
        val permissionAccessFineLocation = ActivityCompat.checkSelfPermission(
            this, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED

        val locationEnabled = requestingLocationUpdates
                && (permissionAccessCoarseLocation || permissionAccessFineLocation)

        // Get the arguments from the intent from MainActivity.
        val intent = intent
        val roll = intent.getParcelableExtra<Roll>(ExtraKeys.ROLL)
        if (roll == null) finish()

        // The point at which super.onCreate() is called is important.
        // Calling it at the end of the method resulted in the back button not appearing
        // when action mode was enabled.
        super.onCreate(savedInstanceState)

        // If the device is locked, this activity can be shown regardless.
        // This way the user doesn't have to unlock the device with authentication
        // just to access this activity.
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1) {
            setShowWhenLocked(true)
            setTurnScreenOn(true)
        } else {
            val window = window
            @Suppress("DEPRECATION") window.addFlags(WindowManager.LayoutParams.FLAG_DISMISS_KEYGUARD)
            @Suppress("DEPRECATION") window.addFlags(WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED)
            @Suppress("DEPRECATION") window.addFlags(WindowManager.LayoutParams.FLAG_TURN_SCREEN_ON)
        }

        // Use the same activity layout as in MainActivity.
        setContentView(R.layout.activity_main)
        val container = findViewById<FrameLayout>(R.id.fragment_container)
        container.transitionName = "roll_transition_${roll?.id}"

        if (findViewById<View?>(R.id.fragment_container) != null && savedInstanceState == null) {

            // Pass the arguments from MainActivity on to FramesFragment.
            val framesFragment = FramesFragment()
            val arguments = Bundle()
            arguments.putParcelable(ExtraKeys.ROLL, roll)
            arguments.putBoolean(ExtraKeys.LOCATION_ENABLED, locationEnabled)
            framesFragment.arguments = arguments
            supportFragmentManager.beginTransaction().add(R.id.fragment_container,
                    framesFragment, FramesFragment.FRAMES_FRAGMENT_TAG).commit()
        }
    }

}